feat: Add `prometheus.echo` component for local metrics inspection (#4105)
grafana/alloy: docs/sources/reference/components/prometheus/prometheus.echo.md

---
canonical: https://grafana.com/docs/alloy/latest/reference/components/prometheus/prometheus.echo/
aliases:
  - ../prometheus.echo/ # /docs/alloy/latest/reference/components/prometheus.echo/
description: Learn about prometheus.echo
labels:
  stage: general-availability
  products:

>     - oss

Reviewer: I don't think this needs to be a community component, the Alloy team should be able to maintain it moving forward.

title: prometheus.echo
---

> # `prometheus.echo`
> 

Reviewer 2: ```suggestion
# `prometheus.echo`

{{< docs/shared lookup="stability/community.md" source="alloy" version="<ALLOY_VERSION>" >}}

```

Add statement about community supported component.

> The `prometheus.echo` component receives Prometheus metrics and writes them to stdout in Prometheus exposition format.

Reviewer 2: @dehaansa is stdout in this context the same as `level.Info(a.logger)` (line 267 in `echo.go`)? Do we call this stdout? or is this information sent to the Alloy logs? The existing `loki.eco` docs mention sending info to stdout as well, so this might be OK here. Just wanting to make sure.

Reviewer: Alloy logs to stdout at this point in time, then that stdout is captured by the relevant service architecture (windows event logs/systemd+journald/etc) so it seems fine for now, though I believe Joe is investigating adding options to log to file right now. Probably worth updating them both together in the future, should be fine for this PR.

## Blocks

The `prometheus.echo` component doesn't support any blocks. You can configure this component with arguments.

## Exported fields

The following fields are exported and can be referenced by other components:

| Name       | Type                    | Description                              |
| ---------- | ----------------------- | ---------------------------------------- |
| `receiver` | `prometheus.Appendable` | A value that other components can use to send metrics to. |

## Component health

`prometheus.echo` is only reported as unhealthy if given an invalid configuration.
In those cases, exported fields retain their last healthy values.

## Debug information

`prometheus.echo` doesn't expose any component-specific debug information.

## Debug metrics

`prometheus.echo` doesn't expose any component-specific debug metrics.

## Example

This example creates a metrics generation and inspection pipeline:

```alloy
prometheus.exporter.unix "default" {
}

prometheus.scrape "demo" {
  targets    = prometheus.exporter.unix.default.targets
  forward_to = [prometheus.echo.debug.receiver]
}

prometheus.echo "debug" {
  format = "text"
}
```

In this example:

> 1. The `prometheus.exporter.unix` component exposes system metrics.
> 2. The `prometheus.scrape` component scrapes those metrics.
> 3. The `prometheus.echo` component receives the scraped metrics and writes them to stdout in Prometheus text format.

Reviewer 2: ```suggestion
1. The `prometheus.exporter.unix` component exposes system metrics.
1. The `prometheus.scrape` component scrapes those metrics.
1. The `prometheus.echo` component receives the scraped metrics and writes them to stdout in Prometheus text format.
```

We use "lazy numbering" in the docs.

> When you run this configuration, you'll see the metrics being written to Alloy's stdout, which is useful for:

Reviewer 2: ```suggestion
When you run this configuration, you'll see the metrics being written to stdout, which is useful to:
```

We shouldn't use possessive "Alloy's" in the docs. I think it's probably safe to drop it here given the context.

> * Debugging metric collection issues
> * Verifying metric labels and values
> * Testing metric transformations
> * Understanding the structure of metrics in your pipeline

Reviewer 2: ```suggestion
* Debug metric collection issues
* Verify metric labels and values
* Test metric transformations
* Understand the structure of metrics in your pipeline
```

We avoid gerunds in the docs (words ending in "ing"). A small tweak to the sentence leading to the list and we can from the "ing"

### Example with OpenMetrics format

```alloy
prometheus.scrape "demo" {
  targets = [
    {"__address__" = "localhost:9090"},
  ]
  forward_to = [prometheus.echo.debug.receiver]
}

prometheus.echo "debug" {
  format = "openmetrics"
}
```

This example outputs metrics using the OpenMetrics format instead of the traditional Prometheus text format.

<!-- START GENERATED COMPATIBLE COMPONENTS -->

## Compatible components

`prometheus.echo` can accept arguments from the following components:

> - Components that export [Targets](../../../compatibility/#targets-exporters)
> 
> 
> ## Has exports

Reviewer 2: ```suggestion
- Components that export [Targets](../../../compatibility/#targets-exporters)

## Has exports
```

Removing empty line
